### DISCUSSION TITLE
Request help on coding Arduino UNO Rev 2 Bluetooth and DRV8825 for Stepper Motor

### DISCUSSION THREAD

**Discussion by traveller17** [community/community#44622] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

Request help on coding Arduino UNO Rev 2 Bluetooth and DRV8825 for Stepper Motor.

Was looking for new entries on Github Forum, but could not find it.
Anyone have experience doing this?
I have tried support at Arduino and Blynk, but so far have not received any assistance.
Attached is my file in Word.
any help would be really appreciated!
[Coding as of 17.docx](https://github.com/community/community/files/10440054/Coding.as.of.17.docx)

**Comment by nvspavankalyanch** [comment]
It looks like you are trying to control a stepper motor using an Arduino UNO Rev 2 and a DRV8825 driver via Bluetooth using the Blynk library.

Here are a few things to consider when working on this project:

- Pin connections: Make sure that the step pin and direction pin of the DRV8825 driver are connected to the correct digital pins on the Arduino UNO Rev 2.

- Power supply: Make sure that the stepper motor and the DRV8825 driver are powered by the correct voltage and current. The DRV8825 driver requires a minimum voltage of 8V and a maximum voltage of 45V.

- Blynk setup: Double-check that the Blynk authentication token, device name, and template ID are correct in the code. Also, make sure that the virtual pins on the Blynk app are properly configured to match the virtual pins in the code.

- Timing: The Tunning function in the code uses delayMicroseconds() to control the speed of the stepper motor. Be careful when adjusting the delay time, as it can affect the precision of the stepper motor.

- Debugging: If you encounter any errors, use the Serial.print() function to debug your code and check the values of variables in real-time.

- While the code you've provided looks like it is mostly correct, there are a few syntax errors that need to be corrected in order for the code to compile.
- It would be helpful to review the code and correct the syntax errors, also make sure that the libraries you are using are installed correctly and that you are using the correct versions of them.


Here are a few additional things to consider:

- **Syntax errors**: There are a few syntax errors in the code that you've provided. For example, the virtualWrite function calls in the setup() function are not written correctly. The correct syntax is:

```
BLYNK_WRITE(V1) {
// code here
}

BLYNK_WRITE(V2) {
// code here
}

BLYNK_WRITE(V3) {
// code here
}

BLYNK_WRITE(V4) {
// code here
}
```

- **Tunning function**: The code you've provided is missing a function called "Tunning()". This function is used to control the stepper motor, and it should be defined before the loop() function. Make sure that the function takes in two parameters, one for the delay time and one for the number of steps.

- **Stepper Motor library**: The code you've provided does not include any library for controlling the stepper motor. You will need to include an appropriate library for the DRV8825 driver, such as the AccelStepper library.

- Make sure that the libraries you are using are installed correctly and that you are using the correct versions of them.

- **Fine tuning**: The FineTuneRight and FineTuneLeft variables are not being set anywhere in the code. You should add lines of code to set these variables based on the values coming from the Blynk app.

- **Serial debugging**: It's a good practice to add Serial.println() statements to check the values of variables during debugging.


Here is an example of how you can modify your code to address the issues I mentioned earlier:


```
#include <Blynk.h>
#include <AccelStepper.h>

#define BLYNK_TEMPLATE_ID "TMPLXrgp_vpa"
#define BLYNK_DEVICE_NAME "TMPLzDfCV5Zh"
#define BLYNK_AUTH_TOKEN "2HPTwd_t80m7lfOB5KTrhWxP7-4JkcqO"

#define BLYNK_PRINT Serial

// Define pin connections
const int stepPin = 3;  // UNO Rev 2 digital pin 3
const int dirPin = 2;  // UNO Rev 2 digital pin 2
const int  steps_per_rev = 200;

// Define Blynk ports
bool TuneRight = false;
bool TuneLeft = false;
bool FineTuneRight = false;
bool FineTuneLeft = false;

// Define networking parameters
char auth[] = "2HPTwd_t80m7lfOB5KTrhWxP7-4JkcqO";  // set the Blynk authentication code

// Create an instance of the AccelStepper class
AccelStepper stepper(1, stepPin, dirPin);

void setup(){
Serial.begin(115200); // baudrate for serial communication
Blynk.begin(auth);

// Declare pins as Outputs
pinMode(stepPin, OUTPUT); // step pin
pinMode(dirPin, OUTPUT); // direction pin

// Set the number of steps per revolution for the stepper motor
stepper.setMaxSpeed(3000);
stepper.setAcceleration(320);

// Set up Blynk virtual pin callbacks
BLYNK_WRITE(V1) {
TuneLeft = param.asInt(); // turn left
}

BLYNK_WRITE(V2) {
TuneRight = param.asInt(); // turn right
}

BLYNK_WRITE(V3) {
FineTuneLeft = param.asInt(); // fine tuning Left
}

BLYNK_WRITE(V4) {
FineTuneRight = param.asInt(); // fine tuning Right
}
}

void loop() {

Blynk.run();

if (TuneRight){ // turn right
stepper.setSpeed(3000);
stepper.move(320);
}

if (FineTuneRight){ // fine turn right
stepper.setSpeed(16);
stepper.move(10000);
}

if (TuneLeft){ // turn left
stepper.setSpeed(-3000);
stepper.move(-320);
}

if (FineTuneLeft){ // fine turn left
stepper.setSpeed(-16);
stepper.move(-10000);
}

// Run the stepper motor
stepper.run();
}
```



Here, I've included the AccelStepper library, which provides a simple interface for controlling stepper motors. I have created an instance of the AccelStepper class in the global scope and set the number of steps per revolution for the stepper motor. In the setup() function, I've set the maximum speed and acceleration for the stepper motor.

In the loop() function, I've added code to check the values of the TuneRight, TuneLeft, FineTuneRight and FineTuneLeft variables and use the stepper.move() function to control the stepper motor based on these values. The stepper.move() function takes in the number of steps as a parameter. I've also added the code to run the stepper motor.

I've also corrected the syntax errors in the BLYNK_WRITE calls, and added the missing parenthesis.

**Note: This code is an example and you may need to adjust the parameters and values to suit your specific application.** 

**I hope this information helps you control your stepper motor successfully. If you have any further issues, feel free to ask for help in the Arduino or Blynk community forums**

**Comment by traveller17** [comment]
Thank you very much for your detailed answer.
I will review tomorrow...

Sent from my iPad,
Regards Lawrence

On Jan 17, 2023, at 11:52 PM, N V S Pavan Kalyan ***@***.***> wrote:

﻿
It looks like you are trying to control a stepper motor using an Arduino UNO Rev 2 and a DRV8825 driver via Bluetooth using the Blynk library.

Here are a few things to consider when working on this project:

Pin connections: Make sure that the step pin and direction pin of the DRV8825 driver are connected to the correct digital pins on the Arduino UNO Rev 2.

Power supply: Make sure that the stepper motor and the DRV8825 driver are powered by the correct voltage and current. The DRV8825 driver requires a minimum voltage of 8V and a maximum voltage of 45V.

Blynk setup: Double-check that the Blynk authentication token, device name, and template ID are correct in the code. Also, make sure that the virtual pins on the Blynk app are properly configured to match the virtual pins in the code.

Timing: The Tunning function in the code uses delayMicroseconds() to control the speed of the stepper motor. Be careful when adjusting the delay time, as it can affect the precision of the stepper motor.

Debugging: If you encounter any errors, use the Serial.print() function to debug your code and check the values of variables in real-time.

While the code you've provided looks like it is mostly correct, there are a few syntax errors that need to be corrected in order for the code to compile.

It would be helpful to review the code and correct the syntax errors, also make sure that the libraries you are using are installed correctly and that you are using the correct versions of them.

Here are a few additional things to consider:

Syntax errors: There are a few syntax errors in the code that you've provided. For example, the virtualWrite function calls in the setup() function are not written correctly. The correct syntax is:
BLYNK_WRITE(V1) {
// code here
}

BLYNK_WRITE(V2) {
// code here
}

BLYNK_WRITE(V3) {
// code here
}

BLYNK_WRITE(V4) {
// code here
}
Tunning function: The code you've provided is missing a function called "Tunning()". This function is used to control the stepper motor, and it should be defined before the loop() function. Make sure that the function takes in two parameters, one for the delay time and one for the number of steps.

Stepper Motor library: The code you've provided does not include any library for controlling the stepper motor. You will need to include an appropriate library for the DRV8825 driver, such as the AccelStepper library.

Make sure that the libraries you are using are installed correctly and that you are using the correct versions of them.

Fine tuning: The FineTuneRight and FineTuneLeft variables are not being set anywhere in the code. You should add lines of code to set these variables based on the values coming from the Blynk app.

Serial debugging: It's a good practice to add Serial.println() statements to check the values of variables during debugging.

Here is an example of how you can modify your code to address the issues I mentioned earlier:

#include <Blynk.h>
#include <AccelStepper.h>

#define BLYNK_TEMPLATE_ID "TMPLXrgp_vpa"
#define BLYNK_DEVICE_NAME "TMPLzDfCV5Zh"
#define BLYNK_AUTH_TOKEN "2HPTwd_t80m7lfOB5KTrhWxP7-4JkcqO"

#define BLYNK_PRINT Serial

// Define pin connections
const int stepPin = 3;  // UNO Rev 2 digital pin 3
const int dirPin = 2;  // UNO Rev 2 digital pin 2
const int  steps_per_rev = 200;

// Define Blynk ports
bool TuneRight = false;
bool TuneLeft = false;
bool FineTuneRight = false;
bool FineTuneLeft = false;

// Define networking parameters
char auth[] = "2HPTwd_t80m7lfOB5KTrhWxP7-4JkcqO";  // set the Blynk authentication code

// Create an instance of the AccelStepper class
AccelStepper stepper(1, stepPin, dirPin);

void setup(){
Serial.begin(115200); // baudrate for serial communication
Blynk.begin(auth);

// Declare pins as Outputs
pinMode(stepPin, OUTPUT); // step pin
pinMode(dirPin, OUTPUT); // direction pin

// Set the number of steps per revolution for the stepper motor
stepper.setMaxSpeed(3000);
stepper.setAcceleration(320);

// Set up Blynk virtual pin callbacks
BLYNK_WRITE(V1) {
TuneLeft = param.asInt(); // turn left
}

BLYNK_WRITE(V2) {
TuneRight = param.asInt(); // turn right
}

BLYNK_WRITE(V3) {
FineTuneLeft = param.asInt(); // fine tuning Left
}

BLYNK_WRITE(V4) {
FineTuneRight = param.asInt(); // fine tuning Right
}
}

void loop() {

Blynk.run();

if (TuneRight){ // turn right
stepper.setSpeed(3000);
stepper.move(320);
}

if (FineTuneRight){ // fine turn right
stepper.setSpeed(16);
stepper.move(10000);
}

if (TuneLeft){ // turn left
stepper.setSpeed(-3000);
stepper.move(-320);
}

if (FineTuneLeft){ // fine turn left
stepper.setSpeed(-16);
stepper.move(-10000);
}

// Run the stepper motor
stepper.run();
}
Here, I've included the AccelStepper library, which provides a simple interface for controlling stepper motors. I have created an instance of the AccelStepper class in the global scope and set the number of steps per revolution for the stepper motor. In the setup() function, I've set the maximum speed and acceleration for the stepper motor.

In the loop() function, I've added code to check the values of the TuneRight, TuneLeft, FineTuneRight and FineTuneLeft variables and use the stepper.move() function to control the stepper motor based on these values. The stepper.move() function takes in the number of steps as a parameter. I've also added the code to run the stepper motor.

I've also corrected the syntax errors in the BLYNK_WRITE calls, and added the missing parenthesis.

Note: This code is an example and you may need to adjust the parameters and values to suit your specific application.

I hope this information helps you control your stepper motor successfully. If you have any further issues, feel free to ask for help in the Arduino or Blynk community forums

**Comment by traveller17** [comment]
Hi Pavan,

 

Again thank you so much for your detailed answer.

 

I have gone and checked the items below (in Red) and have tried to compile, but got a  lot of errors. I did it twice, once with initial changes and 2nd time after a few more.

 

If you don’t mind, could you please assist me, as I seem to be digging myself deeper and deeper into a hole, and I am not sure how to get out.

 

Regards,

 

Lawrence

From my W10 PC
